docs/_posts/2022-08-27-zero-inference.md

```diff
@@ -0,0 +1,100 @@
+
+### Models
+We use two Open Pre-trained Transformers (OPT) models: OPT-30B and OPT-175B having 30 and 175 billion parameters respectively. We configure these models for half-precision inference, which requires about 60GB and 328GB of memory for OPT-30B and OPT-175B respectively. ZeRo-Inference is required to inference these models on a single V100-32GB since they are bigger than GPU memory.
+
+### Full Offload vs. Partial Offload of model weights
+A key design choice in ZeRO-Offload is to offload all the weights of models larger than GPU memory rather than host a subset of the weights in GPU memory. Our intuition for this approach is that for throughput-oriented inference applications, the larger batch sizes enabled by full offload yields better performance than partial offload. In *Table 1*, we present results for OPT-30B token generation on a single V100-32GB that compare fully offloading the model weights versus hosting a portion (i.e., 10 and 12 billion parameters[^partial_offload]) in GPU memory. The results show that full offload delivers the best performance for both CPU memory (43 tokens per second) and NVMe memory (30 tokens per second).  With both CPU and NVMe memory, full offload is over 1.3x and 2.4x faster than partial offload of 18 and 20 billion parameters respectively. The performance advantage of full offload comes from the larger batch sizes compared to the partial offload options. **Thus when a model does not fit in GPU, using GPU memory to increase batch size rather than to partially fit the model leads to faster token generation.**
+
+![Full-offload](/assets/images/zero_inference_full_offload.png){: .align-center}
+
+[^partial_offload]: Pinning more parameters in GPU memory resulted in out of memory errors for small batch sizes.
+
+### Prefetching layer weights
+ZeRO-Inference fetches layers ahead of use, overlapping with current layer computation, to hide layer transfer latency. We measure the impact of prefetching on token generation performance for both OPT-30B and OPT-175B models on a single V100-32GB and summarize the results in *Table 2*. We observe that prefetching did not improve CPU offload. This is because the relatively short sequences in token generation (i.e., less than 50 tokens) resulted in layer computation time that is insufficient to hide a significant portion of layer fetch time from CPU.  In contrast, prefetching improves NVMe offloading performance by 1.13x and 1.14x for OPT-30B and OPT-175B respectively.  This is because transferring weights from NVMe through CPU memory allows prefetching to overlap transfers from CPU to GPU memory with transfers from NVMe to CPU boosting the effective transfer bandwidth.
```

Comment:
This is fair point. Since we are measuring token generation/second, the number of generated tokens is irrelevant since each token requires a full forward pass anyways.

Reply:
it does make a difference to the throughput. Due to previous token caching the longer the sequence then less of a batch size one can fit and thus the lower the throughput.

Reply:
Done